internal/ls/completions.go

```diff
@@ -0,0 +1,2306 @@
+		sortedEntries = getJSCompletionEntries(
+			file,
+			position,
+			uniqueNames,
+			compilerOptions.GetEmitScriptTarget(),
+			sortedEntries,
+		)
+	}
+
+	// !!! exhaustive case completions
+
+	var defaultCommitCharacters *[]string
+	if supportsDefaultCommitCharacters(clientOptions) && ptrIsTrue(clientOptions.CompletionItem.CommitCharactersSupport) {
+		defaultCommitCharacters = &data.defaultCommitCharacters
+	}
+
+	var itemDefaults *lsproto.CompletionItemDefaults
+	if defaultCommitCharacters != nil {
+		itemDefaults = &lsproto.CompletionItemDefaults{
+			CommitCharacters: defaultCommitCharacters,
+		}
+	}
+
+	// !!! port behavior of other strada fields of CompletionInfo that are non-LSP
+	return &lsproto.CompletionList{
+		IsIncomplete: data.hasUnresolvedAutoImports,
+		ItemDefaults: itemDefaults,
+		Items:        sortedEntries,
+	}
+}
+
+func (l *LanguageService) getCompletionEntriesFromSymbols(
+	data *completionData,
+	replacementToken *ast.Node,
+	position int,
+	file *ast.SourceFile,
+	program *compiler.Program,
+	target core.ScriptTarget,
+	preferences *UserPreferences,
+	compilerOptions *core.CompilerOptions,
+	clientOptions *lsproto.CompletionClientCapabilities,
+) (uniqueNames *core.Set[string], sortedEntries []*lsproto.CompletionItem) {
+	closestSymbolDeclaration := getClosestSymbolDeclaration(data.contextToken, data.location)
+	useSemicolons := probablyUsesSemicolons(file)
+	typeChecker := program.GetTypeChecker()
+	// Tracks unique names.
+	// Value is set to false for global variables or completions from external module exports, because we can have multiple of those;
+	// true otherwise. Based on the order we add things we will always see locals first, then globals, then module exports.
+	// So adding a completion for a local will prevent us from adding completions for external module exports sharing the same name.
+	uniques := make(uniqueNamesMap)
+	for _, symbol := range data.symbols {
+		symbolId := ast.GetSymbolId(symbol)
+		origin := data.symbolToOriginInfoMap[symbolId]
+		name, needsConvertPropertyAccess := getCompletionEntryDisplayNameForSymbol(
+			symbol,
+			target,
+			origin,
+			data.completionKind,
+			data.isJsxIdentifierExpected,
+		)
+		if name == "" ||
+			uniques[name] && (origin == nil || !originIsObjectLiteralMethod(origin)) ||
+			data.completionKind == CompletionKindGlobal &&
+				!shouldIncludeSymbol(symbol, data, closestSymbolDeclaration, file, typeChecker, compilerOptions) {
+			continue
+		}
+
+		// When in a value location in a JS file, ignore symbols that definitely seem to be type-only.
+		if !data.isTypeOnlyLocation && ast.IsSourceFileJS(file) && symbolAppearsToBeTypeOnly(symbol, typeChecker) {
+			continue
+		}
+
+		originalSortText := data.symbolToSortTextMap[ast.GetSymbolId(symbol)]
+		if originalSortText == "" {
+			originalSortText = SortTextLocationPriority
+		}
+		sortText := core.IfElse(isDeprecated(symbol, typeChecker), deprecateSortText(originalSortText), originalSortText)
```

Comment:
Nit but, this will always evaluate `deprecateSortText`, which you may not want to do?